### PR TITLE
Refactor gen_variable_type

### PR DESCRIPTION
**Pull request by colesbury**
The gen_variable_type.py script now is only responsible for generating
VariableType.h/cpp. The parent script, "gen_autograd.py", delegates to
gen_autograd_functions.py, gen_variable_type.py, and
gen_python_functions.py.

I've removed "fallthrough" functions. It's replaced by
DONT_RECORD_TRACE, DONT_PROFILE, and DONT_REQUIRE_DERIVATIVE.

In preparation for binding the _out variants, I changed some static
types to Tensor (from Variable) and we now unpack and name tuple return
values.

[VariableType.cpp](https://gist.github.com/colesbury/f073eb601b2d95859244c724f0bb11c4)
[VariableType.cpp.diff](https://gist.github.com/colesbury/95dc718b20a2323f087477b0e9aa3481)

### CODE REVIEW

**Comment by colesbury** [comment]
Previously, we did not profile "fall through functions". Many of these functions are now profiled, with the exception of things like size/stride/storage_offset.

Previously, we did not bind functions that were missing derivative definitions. We now generate bindings (and throw not implemented exceptions in backwards). This should make it easier to write and test native functions before writing the derivative formula.

Previously, we had derivative definitions for things like `<`, `>`, `==`. These are now limited to the in-place versions, since the out-of-place versions return BoolTensor (ByteTensor), which is never differentiable. The corresponding tests in `test_autograd.py` are deleted because they otherwise throw "output does not require grad".

**Comment by ezyang** [comment]
The indentation of the new autogenerated code looks slightly botched:

```
@@ -4501,48 +4518,45 @@
     auto& self_ = unpack(self, "self", 0);
     auto& vec2_ = unpack(vec2, "vec2", 1);
     std::shared_ptr<GerBackward> grad_fn;
-    auto requires_grad = compute_requires_grad({ self, vec2 });
-    if (requires_grad) {
+  if (compute_requires_grad({ self, vec2 })) {
       grad_fn = std::make_shared<GerBackward>();
       grad_fn->next_functions = compute_next_functions({ self, vec2 });
       grad_fn->vec2_ = SavedVariable(vec2, false);
       grad_fn->self_ = SavedVariable(self, false);
     }
-    auto ret = as_variable(baseType->ger(self_, vec2_));
-    set_history(ret, grad_fn);
+  auto result = as_variable(baseType->ger(self_, vec2_));
+  set_history(result, grad_fn);
     if (jit::tracer::isTracing( self, vec2 )) {
-      jit::Node *n = jit::tracer::recordTrace( "ger", { self, vec2 }, { ret } );
+    jit::Node *n = jit::tracer::recordTrace( "ger", { self, vec2 }, { result } );
       (void)n;
     }
-    return Tensor(std::move(ret));
+  return result;
 }
```

You're losing a level of indentation at the first if statement.

File: test/test_autograd.py [hunk 1]
```diff
@@ -2236,36 +2236,6 @@ class dont_convert(tuple):
     ('det', lambda: random_fullrank_matrix_distinct_singular_value(S), (), 'distinct_postive_s', (), [skipIfNoLapack]),
     ('svd', lambda: random_fullrank_matrix_distinct_singular_value(S), (), '', (), [skipIfNoLapack]),
     ('gesv', (S, S), ((S, S),), '', (), [skipIfNoLapack]),
-    ('eq', (S, S, S), ((S, S, S),)),
```
Comment (ezyang) [inline test/test_autograd.py]: How did you check that these methods are still being tested?

**Comment by ezyang** [comment]
> The corresponding tests in test_autograd.py are deleted because they otherwise throw "output does not require grad".

I'm confused. If out-of-place comparison test never gave a differentiable output, how was the test passing before? Is this actually another case of #4422?

File: tools/autograd/templates/VariableType.cpp [hunk 2]
```diff
@@ -311,30 +302,33 @@ static void rebase_history(Variable& var, std::shared_ptr<Function> grad_fn, int
...
       var.get()->output_nr = output_nr;
       var.get()->_grad_fn = grad_fn;
       output_nr++;
     }
   }
 }
 
-static variable_list flatten(const TensorList& tensors) {
-  return cast_tensor_list(tensors);
+static at::ArrayRef<Variable> flatten(TensorList tensors) {
```
Comment (ezyang) [inline tools/autograd/templates/VariableType.cpp]: This micro-optimization makes me nervous, as none of the overloads return (nor can they return) an ArrayRef...

**Comment by ezyang** [comment]
The Tensor<->Variable switches make me soooo nervous, but I am told this is all going away when we are merging Tensor and Variable, right? :)

File: tools/autograd/gen_variable_type.py [hunk 3]
```diff
@@ -15,106 +16,96 @@
...
-import re
-import yaml
-from collections import defaultdict
-from .utils import CodeTemplate, write, nested_dict, YamlLoader
+# These functions are not worth profiling because they are very cheap and may
+# be called very often.
+DONT_PROFILE = {
+    'data_ptr', 'get_device', 'is_contiguous', 'is_cuda', 'is_distributed',
+    'is_same_size', 'is_set_to', 'is_signed', 'is_sparse', 'numel',
+    'size', 'storage_offset', 'stride',
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py]: Nice!

File: tools/autograd/gen_variable_type.py [hunk 4]
```diff
@@ -15,106 +16,96 @@
...
+}
 
 METHOD_DECLARATION = CodeTemplate("""\
 virtual ${return_type} ${method_prefix_derived}${api_name}(${formals}) const override;
 """)
 
 METHOD_DEFINITION = CodeTemplate("""\
 ${return_type} VariableType::${method_prefix_derived}${api_name}(${formals}) const {
-    ${type_definition_body}
+  ${type_definition_body}
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py]: ...probably the cause for the indentation problems?

**Review by ezyang — APPROVED** [review]

File: tools/autograd/templates/VariableType.cpp (hunk 2 above)
Comment (apaszke) [reply tools/autograd/templates/VariableType.cpp]: It's not even valid. `tensors` will be a temporary copy of an argument, so the data of this array ref will be freed when returning.

Comment (ezyang) [reply tools/autograd/templates/VariableType.cpp]: Oh yeah, I remember; the output is live only as long as the input is live, and that's a nontrivial invariant to remember especially since it would only apply to this argument.

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -272,12 +262,10 @@
...
   A: -at::mm(std::get<0>(gesv(grad, A.t())), solution.t())
 
-# get_device
-
-- name: gt(Tensor self, Scalar other)
+- name: gt_(Tensor self, Scalar other)
   self: zeros_like(self)
 
-- name: gt(Tensor self, Tensor other)
+- name: gt_(Tensor self, Tensor other)
```
Comment (apaszke) [inline tools/autograd/derivatives.yaml]: Why do we need these specs? If an output never requires grad then the formulas don't matter anyway, right?

File: tools/autograd/templates/VariableType.cpp [hunk 6]
```diff
@@ -345,7 +339,7 @@ static variable_list flatten(const Tensor& x, const TensorList& y) {
   return r;
 }
 
-static variable_list flatten(const Tensor& x, const TensorList& y, const Tensor& z) {
+static variable_list flatten(const Tensor& x, TensorList y, const Tensor& z) {
```
Comment (apaszke) [inline tools/autograd/templates/VariableType.cpp]: Why can't it be a `const&` anymore?

File: tools/autograd/gen_variable_type.py [hunk 7]
```diff
@@ -15,106 +16,96 @@
...
-from .utils import CodeTemplate, write, nested_dict, YamlLoader
+# These functions are not worth profiling because they are very cheap and may
+# be called very often.
+DONT_PROFILE = {
+    'data_ptr', 'get_device', 'is_contiguous', 'is_cuda', 'is_distributed',
+    'is_same_size', 'is_set_to', 'is_signed', 'is_sparse', 'numel',
+    'size', 'storage_offset', 'stride',
+}
 
+DONT_REQUIRE_DERIVATIVE = {
```
Comment (apaszke) [inline tools/autograd/gen_variable_type.py]: can you add a comment saying what happens to those functions (do we still generate them? is their backward just an error? are they guaranteed to return things that don't require grad?)?

File: test/test_autograd.py (hunk 1 above)
Comment (gchanan) [reply test/test_autograd.py]: but these did test some useful stuff like that inplace worked; we are just throwing that away?  Can't we just skip the specific part of the check that triggers the error if the return type of the function is non-differentiable?

**Comment by colesbury** [comment]
RE: indentation: `diff` isn't showing white-space only changes. I changed the generated methods to two-space indent to match the rest of C++ including the non-generated parts of VariableType.cpp.

(See the full VariableType.cpp linked above to verify)

File: tools/autograd/derivatives.yaml (hunk 5 above)
Comment (colesbury) [reply tools/autograd/derivatives.yaml]: Out-of-place `gt` automatically doesn't require grad because the return value is an IndexTensor. However, in-place `gt_` will return a `FloatTensor` if the input is a `FloatTensor`.

File: tools/autograd/gen_variable_type.py (hunk 4 above)
Comment (colesbury) [reply tools/autograd/gen_variable_type.py]: Indentation fixes! 😄 

Here's the full generated code: [VariableType.cpp](https://gist.github.com/colesbury/f073eb601b2d95859244c724f0bb11c4).

File: tools/autograd/templates/VariableType.cpp (hunk 6 above)
Comment (colesbury) [reply tools/autograd/templates/VariableType.cpp]: ArrayRefs (TensorList, IntList) should be passed by value. From LLVM's ArrayRef docs, which we copied into ArrayRef.h:

```
This is intended to be trivially copyable, so it should be passed by value
```

File: test/test_autograd.py (hunk 1 above)
Comment (colesbury) [reply test/test_autograd.py]: We should fix this, but I'm not sure how yet. I don't think we should ignore outputs that don't require_grad. I think we should be checking that the gradient is actually zero. Otherwise we miss things that should have gradient definitions, but are erroneously marked as not requiring grad.

File: tools/autograd/gen_variable_type.py [hunk 8]
```diff
@@ -418,380 +319,184 @@ def emit_record_trace(env, declaration):
...
-        elif declaration['return_type'] == 'std::vector<Tensor>':
-            env['return_value'] = 'ret'
-            env['trace_outputs'] = 'cast_tensor_list(ret)'
-        else:
-            env['return_value'] = '{}(std::move(ret))'.format(declaration['return_type'])
-            env['trace_outputs'] = get_trace_outputs(declaration)
-
-        if declaration['name'] in DONT_RECORD_TRACE:
-            env['record_trace'] = ''
+    def declare_return_values():
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: nit: I feel like "values" isn't the right name here, maybe "objects"?  When I think of return values I think of e.g. "5" which you don't declare.

File: tools/autograd/derivatives.yaml [hunk 9]
```diff
@@ -405,26 +384,17 @@
   self: grad * other
   other: grad * self
 
-- name: multinomial(Tensor self, int64_t num_samples, bool replacement, Generator generator)
```
[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/derivatives.yaml]: Functions that return IndexTensor or BoolTensor never require grad. There's now a warning that says the derivative definition is ignored, so it's safe to delete.

File: test/test_autograd.py (hunk 1 above)
Comment (gchanan) [reply test/test_autograd.py]: I agree.  I just meant for the purposes of this PR you can leave the useful checks and avoid the error by looking at the return type of the function (not if the function requires grad or not).

File: tools/autograd/gen_variable_type.py [hunk 10]
```diff
@@ -125,160 +116,100 @@
...
+        # assume that only the first output is differentiable. TODO: remove
+        # this heuristic.
+        differentiable_outputs = differentiable_outputs[:1]
 
-def create_variable_type(top_env, aten_declarations):
-    """VariableType.h and VariableType.cpp body
+    requires_derivative = (
+        name not in DONT_REQUIRE_DERIVATIVE and
+        len(differentiable_inputs) > 0 and len(differentiable_outputs) > 0 and
+        strategy == 'use_derived')
```
[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/gen_variable_type.py]: Python `enum` is new in 3.4.

File: tools/autograd/gen_variable_type.py (hunk 8 above)
Comment (colesbury) [reply tools/autograd/gen_variable_type.py]: Changed to "declare_returned_variables"

File: tools/autograd/gen_variable_type.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: You can fake an enum by doing
```python
VALUE1 = object()
VALUE2 = object()
```
then use `is` for comparison (although I think `==` works too)

[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/gen_variable_type.py]: I forgot about that trick, but I don't see how the fake enum is any better than a string without static typing. strings have the advantage of being easy to debug: `print(strategy)` will give something meaningful.

File: test/test_autograd.py (hunk 1 above)
Comment (colesbury) [reply test/test_autograd.py]: I added back the working in-place tests (`broadcast_rhs/all` doesn't work with in-place).

Adding an exclude requires excluding all of `do_test`, so it won't actually do anything.

File: tools/autograd/derivatives.yaml [hunk 11]
```diff
@@ -247,10 +237,10 @@
 - name: gather(Tensor self, int64_t dim, Tensor index)
   self: grad.type().zeros(self.sizes()).scatter_add_(dim, index, grad)
 
-- name: ge(Tensor self, Scalar other)
+- name: ge_(Tensor self, Scalar other)
```
[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/derivatives.yaml]: I'm not a fan of file-level comments that describe what changed since a while ago. They're only relevant to a few people (pretty much the people on this PR) and only for a little while. They're better placed in commit messages.

Everything in Declarations.yaml is bound in `VariableType.cpp` with the exception of `_out` variants, which will be added soon.

`should_generate_python_binding` decides what has Python-level bindings. Some functions are manually excluded by `SKIP_PYTHON_BINDINGS`. We currently don't Python generate bindings for namespace-only (i.e. `at::`) functions, but that will change soon.

`derivatives.yaml` already describes binding of derivatives. It doesn't describe Python bindings or whether a function is overridden in VariableType.cpp because it doesn't affect those things.

File: test/test_autograd.py (hunk 1 above)
Comment (gchanan) [reply test/test_autograd.py]: that seems fine, but there should have already been guards against running broadcast_lhs/broadcast_all tests inplace, so I don't know why you had to do this.

File: tools/autograd/gen_variable_type.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: fake enum is better because if you make a typo, you'll get an unbound local error. if it's a string it will silently compare as different

[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: If you want nicer prints a hack would be to do this:
```
from collections import namedtuple
VALUE1 = namedtuple('Value1', [])()
VALUE2 = namedtuple('Value2', [])()

print(VALUE1) # Value1()
```

[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/gen_variable_type.py]: differently named tuples are equal:

```python
from collections import namedtuple
VALUE1 = namedtuple('Value1', [])()
VALUE2 = namedtuple('Value2', [])()
print(VALUE1 == VALUE2)  # True (!!!)
```

[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: eh ok. maybe this:
```python
def make_const(name):
    return type(name, (object,), {'__repr__': (lambda self: name)})()
VALUE1 = make_const('VALUE1')
```

[the remark this answers is not in the document]
Comment (ezyang) [reply tools/autograd/gen_variable_type.py]: No matter what, it will be really obvious if you typo the string because the generated code will be all wrong ;)

[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: Unless it will be a particularily nasty error that won't trigger type-level errors in C++ 😉
